### PR TITLE
8252730: jlink does not create reproducible builds on different servers

### PR DESCRIPTION
**Pull request by igraves**
Related to [JDK-8252730 jlink does not create reproducible builds on different servers](https://bugs.openjdk.java.net/browse/JDK-8252730). Introduces ordering based on `Archive` module names to ensure stable files (and file signatures) across generated JDK images by jlink.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252730](https://bugs.openjdk.java.net/browse/JDK-8252730): jlink does not create reproducible builds on different servers

### Reviewers
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to 122bbdfc1dadd29563b7c7c4bb7853163eeebbf1
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/156/head:pull/156`
`$ git checkout pull/156`

### CODE REVIEW

**Comment by stuart-marks** [comment]
Sorting the Archive instances makes more sense to me than trying to add hashCode/equals to Archive. I note that the instances added to the ResourcePoolManager are stored in a LinkedHashMap, so the sorted order should be preserved. I'll defer to Alan and Jim as to whether sorting by module name is the right thing. Additional code comments to follow.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java
Comment (stuart-marks) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java:274]: I'm somewhat allergic to nested forEach plus statement lambdas. I note that the original code mixed an imperative for-loop with a stream, which was kind of odd; converting to a straight nested for-loop might be reasonable. Alternatively, the nested forEaches could be converted to a flatMap as follows:

`
        archives.stream()
                .map(Archive::moduleName)
                .sorted()
                .flatMap(mn -> entriesForModule.get(mn).stream()
                               .map(e -> new ArchiveEntryResourcePoolEntry(mn, e.getResourcePoolEntryName(), e)))
                .forEach(resources::add);
`

Comment (igraves) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java:274]: This is a good catch. I'm partial to flatMap, but unsure if that gets too clever what we're trying to do here or not.

**Review by mlchung — APPROVED** [review]
Looks good to me.

**Comment by igraves** [comment]
Updated to merge streams in line with @stuart-marks suggestion.

**Comment by AlanBateman** [comment]
There are several existing jlink tests that check that generated lib/modules is reproducible. Can we do more in these tests or add a new test that would improve the chances of finding issues, e.g. using jlink to create a JDK that includes the jlink tool and the packaged modules, copy it to two locations and run the jlink in both run-time images to see if they produce the same bits.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java
Comment (dfuch) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java:41]: nit: `Comparator` might no longer be needed? I don't see where it's used.

Comment (igraves) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java:41]: Yep. Looks like it's a leftover. Thanks!

**Review by AlanBateman — APPROVED** [review]
The update using flatMap looks good. I think we need to explore adding more tests if possible.

**Comment by AlanBateman** [comment]
Thanks for adding a test. What would you think about adding this to JLinkReproducible2Test to avoid duplication. The existing test can be tweaked to do the mismatch check with different sets of modules. In passing, no need to include java.base in the @modules tag.

File: test/jdk/tools/jlink/JLinkReproducible3Test.java
Comment (AlanBateman) [inline test/jdk/tools/jlink/JLinkReproducible3Test.java:112]: The update to ImageFileCreator looks good.

A few nits in JLinkReproducible3Test:
1. Invoking copyJDKs with two destination locations looks a bit strange. I think would be easier for future maintainers if it does one copy and is called twice.
2. Files.copy will create a directory, no need for Files.createDirectories.
3. Are you sure that bin/jlink is executable, I just wonder if the test runs with the patch have ended up depending the umask at the time. It is possible to specify COPY_ATTRIBUTES to the copy method and it will attempt to copy the permissions.
4. The variable naming is more like C style, so a bit inconsistent with the naming usually used in Java programs.

Comment (igraves) [reply test/jdk/tools/jlink/JLinkReproducible3Test.java:112]: 1. I agree that it's awkward and I'm not a huge fan. I'm okay doing it this way, but this is recreating the reproducer for https://bugs.openjdk.java.net/browse/JDK-8252730 which is two identical JDK's in different directories. I can document this in the test further or we can change it.

2-4: Noted. Will clean these up and loop back.

Comment (AlanBateman) [reply test/jdk/tools/jlink/JLinkReproducible3Test.java:112]: Whatever is easiest but 2 x copy tree would be simpler and the result should be two identical JDKs as reported in JDK-8252730.

Comment (igraves) [reply test/jdk/tools/jlink/JLinkReproducible3Test.java:101]: I misunderstood what you meant when I read it earlier today, apologies! The new changes should address these enumerated issues.

File: test/jdk/tools/jlink/JLinkReproducible3Test.java
Comment (AlanBateman) [inline test/jdk/tools/jlink/JLinkReproducible3Test.java:67]: Could use copyJDk1Dir.resolve("bin").resolve("jlink") here to avoid going to strings and back.

File: test/jdk/tools/jlink/JLinkReproducible3Test.java
Comment (AlanBateman) [reply test/jdk/tools/jlink/JLinkReproducible3Test.java:101]: The updated test looks okay, I can sponsor if you need it.

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**
